File: Source Code/website/lib/models/scorecard.dart
import './committee.dart';

class Parameter {
  late int id;
  String title;
  int maxScore;

  static int _id = 1;

  Parameter(this.title, this.maxScore, {int? id}) {
    if (id != null) {
      this.id = id;

      if (id >= _id) {
        _id = id + 1;
      }
    } else {
      this.id = _id++;
    }
  }

  @override
  String toString() => "$title ($maxScore)";
}

class Scorecard {
  late List<Parameter> parameters;
  late Map<String, List<double>> scores;

  Scorecard(Committee committee) {
    parameters = [
      Parameter("GSL", 10),
      Parameter("Mod", 10),
      Parameter("POI", 10),
      Parameter("Chits", 5),
      Parameter("Total", 0),
    ];

    scores = {
      for (final String delegate in committee.delegates) delegate: [0, 0, 0, 0],
    };
  }

  Scorecard.fromJson(Map<String, dynamic> data) {
    parameters = List.generate(
      data["parameters"].length,
      (index) => Parameter(
        data["parameters"][index],
        data["maxScores"][index],
      ),
    );

    scores = Map<String, List<double>>.from(
      data["scores"].map<String, List<double>>(
        (key, value) =>
            MapEntry<String, List<double>>(key, value.cast<double>()),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "parameters": parameters.map<String>((e) => e.title).toList(),
      "maxScores": parameters.map<int>((e) => e.maxScore).toList(),
      "scores": scores,
    };
  }
}

File: Source Code/react-website/src/types/scorecard.ts
export interface Parameter {
  id: number;
  title: string;
  maxScore: number;
}

let _nextParameterId = 1;

export function createParameter(
  title: string,
  maxScore: number,
  id?: number,
): Parameter {
  if (id !== undefined) {
    if (id >= _nextParameterId) _nextParameterId = id + 1;
    return { id, title, maxScore };
  }
  return { id: _nextParameterId++, title, maxScore };
}

export interface Scorecard {
  parameters: Parameter[];
  scores: Record<string, number[]>;
}

export function createScorecard(delegates: string[]): Scorecard {
  return {
    parameters: [
      createParameter("GSL", 10),
      createParameter("Mod", 10),
      createParameter("POI", 10),
      createParameter("Chits", 5),
      createParameter("Total", 0),
    ],
    scores: Object.fromEntries(delegates.map((d) => [d, [0, 0, 0, 0]])),
  };
}

export function scorecardFromJson(data: Record<string, unknown>): Scorecard {
  const params = (data.parameters as string[]) ?? [];
  const maxScores = (data.maxScores as number[]) ?? [];
  return {
    parameters: params.map((title, i) =>
      createParameter(title, maxScores[i] ?? 0),
    ),
    scores: Object.fromEntries(
      Object.entries(data.scores as Record<string, number[]>).map(([k, v]) => [
        k,
        [...v],
      ]),
    ),
  };
}

export function scorecardToJson(sc: Scorecard): Record<string, unknown> {
  return {
    parameters: sc.parameters.map((p) => p.title),
    maxScores: sc.parameters.map((p) => p.maxScore),
    scores: sc.scores,
  };
}
